internal/jobobject/iocp.go

```diff
@@ -0,0 +1,83 @@
+package jobobject
+
+import (
+	"context"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	ioInitOnce sync.Once
+	initIOErr  error
+	// Global iocp handle that will be re-used for every job object
+	ioCompletionPort windows.Handle
+	// Mapping of job handle to channel that that job object owns to be able to receive
+	// notifications.
+	jobMap = make(map[uintptr]chan interface{})
+)
+
+// AllProcessesExited is a type representing a notification that every process in a job has exited.
+type AllProcessesExited struct{}
+
+// ioCompletionProcessor processes completed async IOs forever
+func pollIOCP(ctx context.Context, iocpHandle windows.Handle) {
+	var (
+		overlapped uintptr
+		code       uint32
+		key        uintptr
+	)
+
+	for {
+		err := winapi.GetQueuedCompletionStatus(iocpHandle, &code, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("failed to poll for job object notification")
+			continue
+		}
+		if ch, ok := jobMap[key]; ok {
+			// Check code and parse out relevant information related to that notification
+			// that we care about. For now all we handle is the message that all processes
+			// in the job have exited.
+			switch code {
```

Comment:
Actually, I would pull this switch out into another function, then the code can just be:

```
notification, err := parseNotification(code, overlapped)
if err != nil {
    log.G(ctx).WithFields(logrus.Fields{
        "code": code,
        "overlapped": overlapped,
    }).Warn("failed to parse job object notification")
    continue
}
ch <- notification

[...]

func parseNotification(code uint32, overlapped uintptr) (interface{}, error) {
    switch code {
        case winapi.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO:
            return AllProcessesExited{}, nil
        case winapi.JOB_OBJECT_MSG_END_OF_JOB_TIME:
        [...]
        default:
            return nil, fmt.Errorf("unknown job notification type: %d", code)
    }
}
```

This also makes it easier to log different errors if we encounter other parsing problems with more complex notification types in the future.

Reply:
Although, this makes it difficult to distinguish between an error for an unrecognized code, and failing to parse some other part of the notification. Those cases should probably be logged at warn and error level, respectively, so some way to distinguish would be nice.

Reply:
I think when the time comes where we'll actually implement more than just the one we do now we can come up with something